### PR TITLE
[SPARK-2491] Don't handle uncaught exceptions from tasks that fail during executor shutdown.

### PR DESCRIPTION
**Pull request by witgo**

### CODE REVIEW

**Comment by ilganeli** [comment]
@JoshRosen There's been no activity here for over 5 months - is this still alive? If not, I wouldn't mind finishing the job since I ran into this exact issue and could use better debug visibility.

**Comment by JoshRosen** [comment]
@ilganeli I think that this has largely stalled because this is a somewhat complicated-to-review change (as the discussion above shows, it can be tricky to make sure that we've thought through all of the ways in which control can enter the error-handling / shutdown hook code).  This issue is probably still worth fixing, but I think it will take a lot more work to review this, especially since this type of code is hard to test.

**Comment by ilganeli** [comment]
I actually wouldn't mind working on this - it's an issue near and dear to my heart :-). What's the best way to pick up the ball here ? I could just make a new PR - the changes are pretty minimal at present.



Sent with Good (www.good.com)


-----Original Message-----

**Comment by witgo** [comment]
Jenkins, retest this please.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 1]
```diff
@@ -280,13 +280,18 @@ private[spark] class Executor(
...
-          val taskEndReason = new ExceptionFailure(t, metrics)
-          execBackend.statusUpdate(taskId, TaskState.FAILED, ser.serialize(taskEndReason))
-
-          // Don't forcibly exit unless the exception was inherently fatal, to avoid
-          // stopping other tasks unnecessarily.
-          if (Utils.isFatalError(t)) {
-            SparkUncaughtExceptionHandler.uncaughtException(t)
+          if (Utils.inShutdown()) {
+            val reason = ExecutorExitFailure(None, metrics)
+            execBackend.statusUpdate(taskId, TaskState.FAILED, ser.serialize(reason))
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: I guess this doesn't change what was there before, but is this safe during shutdown?

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 2]
```diff
@@ -280,13 +280,18 @@ private[spark] class Executor(
...
+          }
+          else {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: nit: else goes in previous line

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 3]
```diff
@@ -701,6 +701,10 @@ private[spark] class TaskSetManager(
             s"${ef.className} (${ef.description}) [duplicate $dupCount]")
         }
 
+      case e :ExecutorExitFailure =>
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: nit: `case e: ...`

**Comment by vanzin** [comment]
Well, this has been lingering here for a while. Looks sane though.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 1 above)
Comment (witgo) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: Yes,The code work in most cases.

**Comment by andrewor14** [comment]
What's the status of this patch? Is it still relevant after so many releases? @witgo @vanzin

**Comment by witgo** [comment]
I think it is necessary to merge the PR into master.

**Comment by nssalian** [comment]
What is the status of the PR? Seems no movement for a while. @vanzin

**Comment by srowen** [comment]
Do you mind closing this PR?

**Pull request closed without merging**